src/section/section_keys.rs

```diff
@@ -50,12 +65,92 @@ impl SectionKeysProvider {
 
     pub fn finalise_dkg(&mut self, public_key: &bls::PublicKey) {
         if let Some(share) = &self.pending {
-            let pending_public_key = share.public_key_set.public_key();
+            if *public_key != share.public_key_set.public_key() {
+                return;
+            }
+        }
+        if let Some(share) = self.pending.take() {
+            if let Some(evicted) = self.cache.add(public_key, share) {
+                trace!("evicted old key from cache: {:?}", evicted);
+            }
+            trace!("finalised DKG: {:?}", public_key);
+        }
+    }
+}
+
+/// Implementation of simple cache.
+#[derive(Debug)]
+pub struct KeyCache {
+    map: BTreeMap<Vec<u8>, SectionKeyShare>,
+    list: VecDeque<bls::PublicKey>,
+    capacity: usize,
+}
```

Comment:
I'd consider using `LruCache` from the `lru_time_cache` crate instead of rolling your own. We already use it in other places and it works fine.

Reply:
It's just a strange thing to include time here. The impl is more or less the same as that of LruCache, only simpler.

Reply:
The time is optional. `LruCache` can be time-based, capacity-based or both. So in this case you would use capacity only: https://docs.rs/lru_time_cache/0.11.5/lru_time_cache/struct.LruCache.html#method.with_capacity